[iOS] [SaferC++] Improve memory safety in iOS specific WebKit/ code

zakariaridouh opened the pull request:
#### 973527f0dd004747005bef6c8d2499d3d189c8b6
<pre>
[iOS] [SaferC++] Improve memory safety in iOS specific WebKit/ code
<a href="https://bugs.webkit.org/show_bug.cgi?id=302745">https://bugs.webkit.org/show_bug.cgi?id=302745</a>
<a href="https://rdar.apple.com/165001830">rdar://165001830</a>

Reviewed by Vitor Roriz.

Fix failing files for iOS only codepaths.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::providePresentingApplicationPID const):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::dispatchMessage):
(WebKit::NetworkProcessConnection::dispatchSyncMessage):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeConnection):

Canonical link: <a href="https://commits.webkit.org/303874@main">https://commits.webkit.org/303874@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d7464d4babc026cdf1785c35479609b082bce9f7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133612 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6117 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44774 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141173 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85666 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fb87bcf0-a2f0-4b7d-9996-0a5dab2a7cc9) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6710 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5981 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102211 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69571 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e74101dd-759c-4aee-a039-507b094bd0d3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136559 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4742 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119986 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83008 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9d9e2e7f-1924-48e6-b922-0c443721170f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4621 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2188 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113758 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143821 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5789 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38480 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110589 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5869 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4959 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110771 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28141 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4434 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116038 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59538 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5841 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34350 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5689 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69299 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5932 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5796 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -438,7 +438,7 @@ void WebProcess::initializeConnection(IPC::Connection* connection)
     protectedEventDispatcher()->initializeConnection(*connection);
 #if PLATFORM(IOS_FAMILY)
-    m_viewUpdateDispatcher.initializeConnection(*connection);
+    Ref { m_viewUpdateDispatcher }->initializeConnection(*connection);
```
`ViewUpdateDispatcher` implements ref/def, I wonder if we could make m_viewUpdateDispatcher a const ViewUpdateDispatcher and if this would make the static analyzer happy.

vitorroriz replied:
wait, m_viewUpdateDispatcher is not a RefPtr<ViewUpdateDispatcher>, it is a ViewUpdateDispatcher itself 🤔

cdumez replied:
This is a weird but correct case. m_viewUpdateDispatcher is not heap allocated, and it is owned by the WebProcess. ViewUpdateDispatcher is ref countable but doesn't not subclass RefCounted, it merely forwards its ref-counting to its owner (the WebProcess object). So we we ref `m_viewUpdateDispatcher` here, we end up simply ref'ing `this`, which does keep `m_viewUpdateDispatcher` alive, transitively.

It's a good question whether or not marking the data member `const` would silence the static analysis warnings in this case. I'm not sure! (cc @rniwa). Worth trying.

vitorroriz commented on Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
```diff
@@ -72,6 +72,7 @@
 #include <WebCore/NavigationScheduler.h>
 #include <WebCore/Page.h>
 #include <WebCore/PermissionController.h>
+#include <WebCore/ScrollingCoordinator.h>
```
why do we need this now?

zakariaridouh replied:
removed!

zakariaridouh committed:
[iOS] [SaferC++] Improve memory safety in iOS specific WebKit/ code
https://bugs.webkit.org/show_bug.cgi?id=302745
rdar://165001830

Reviewed by Vitor Roriz.

Fix failing files for iOS only codepaths.

* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::providePresentingApplicationPID const):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::dispatchMessage):
(WebKit::NetworkProcessConnection::dispatchSyncMessage):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeConnection):

Canonical link: https://commits.webkit.org/303874@main
